### ISSUE TITLE
Application for Custom Models

### ISSUE THREAD

**Issue by qlawyer** [MaximIntegratedAI/ai8x-synthesis#134]
Hi, 

My ultimate goal is to deploy a custom CNN model, which is available as Pytorch file. YAML file asks for the network structure used in the training and the dataset. Currently it appears to me that I need to redefine my network using ai8x libraries and retrain my model using the tools of maxim so that I can have the required files. Is it the way or am I misreading the documentation? If I (hopefully) misread it, do you have another source/documentation that describes how to use a pre-trainmodel? 

Thank you for the support in advance.

**Comment by rotx-maxim** [comment]
Most likely, you will have to re-train your model.
1. Your model can only use the operators and options supported by hardware (for example, no 5x5 kernels).
2. The hardware uses integer arithmetic, so there is clipping in each layer. Since clipping is a non-linearity, it is important to take that into account during training. When you use the ai8x subclasses (libraries), this is done correctly.
3. While in some cases a model trained for floating point can be quantized after the fact by selecting a scale factor or other method, it's usually much easier to use quantization aware training and let the machine deal with it, and it almost always gives better results.

However, you don't have to use the Maxim tools to re-train your model. All you really need are the subclasses, and you can use your own tools. We have several partners who adopted this into their own custom training flow. However, if your existing pipeline isn't special, it might just be easier to use the training software from our repo (in addition to QAT, we recently added some NAS capabilities, so it does have some useful features).

Hope this explains it a little better!

**Comment by qlawyer** [comment]
Thank you for the explanation. My intention is to avoid defining my network again. It does not matter which tool i use. I need a benchmark for several networks. It means I have to redefine all the network topologies using the ai8x (sub)classes and retrain. I would appreciate if you confirm my understanding. 

Once again, thank you for the assistance.

**Comment by rotx-maxim** [comment]
If your intention is to ultimately run these networks on the MAX78000 hardware, then yes, you need to retrain.

I would consider training for a few epochs only at first and to load the result into the hardware. That way you can ensure you aren't exceeding any hardware limits, that the network is working as expected, and you can measure the power consumption and energy without spending too much time on training. Once it's working, you can refine the training process for decent results with respect to your Loss function (in addition to epoch count, there's always batch size, batch norm, learning rate, optimizer, etc.).

The EVkit has a power monitor built in that will allow you to measure the actual power consumption and energy for an inference (without the power monitor, this could be difficult due to the high dynamic range and the short time for each inference).

**Comment by qlawyer** [comment]
The time consuming part here is to define the network using the ai8x library. In terms of power, you are right. I do not even need the dataset, I can just duplicate a sample instance. I have the EVkit, will surely use this feature. Having looked at the development flow scheme, I thought I can give an arbitrary pytorch or onnx check point and maxim tools handle the rest. Apparently I was too optimistic :).

**Comment by rotx-maxim** [comment]
It depends on what your ultimate goal is. If it's just for benchmarking, the tools should recognize checkpoint parameters for Conv2d, Conv1d, ConvTranspose2d, and Linear. You could then run a quick "naive" quantization on the checkpoint file and convert to C code and run it. Writing the YAML file is still required because the checkpoint does not have that information, and it does not have input data size/channel count or anything about pooling, activation, or elementwise operations. The performance would most likely be pretty questionable (read: bad). Adding in the fusion between pooling and convolution shouldn't really affect the energy too much (but doing so saves hardware resources).

That said, the effort to "port" a model is dependent on how complex your model is. The predefined operators (see "List of Predefined Modules" in README.md) are very similar to PyTorch's modules, and sequential models are very quick to port. More complicated structures (such as skip connections or residuals) will take more time, particularly with regard to resource allocation in the YAML file.
